File: C++/347. TopKFrequentElements.cpp

/* PROBLEM : Given a non-empty array of integers, return the k most frequent elements.

APPROACH: Approached using a map data structure to count the frequency of each element.
*/




class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        map<int,int> m;
        for(auto x:nums) m[x]++;
        vector<pair<int,int>> v;
        for(auto x:m){
            v.push_back({x.second,x.first});
        }
        sort(v.begin(),v.end());
        vector<int> ans;
        int ct=0,pt=v.size()-1;
        while(ct<k && ct<v.size()){
            ans.push_back(v[pt].second);
            pt--;
            ct++;
        }
        return ans;
    }
};

File: Python/347. TopKFrequentElements.py
# 347. Top K Frequent Elements (https://leetcode.com/problems/top-k-frequent-elements/)

class Solution(object):
    def topKFrequent(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """
        frequencies = {}
        for num in nums:
            if num in frequencies:
                frequencies[num] = frequencies[num] + 1
            else:
                frequencies[num] = 1
        
        keys = []
        while len(keys) < k:
            maxx = None
            keyy = None
            for key, value in frequencies.items():
                if not maxx:
                    keyy = key
                    maxx = value
                    continue
                if value > maxx:
                    maxx = value
                    keyy = key
            keys.append(keyy)
            frequencies.pop(keyy)
        return keys
                
